--- index.original.js
+++ index.js
@@ -13,7 +13,6 @@
 
 var http = require('http');
 var https = require('https');
-var PacProxyAgent = require('pac-proxy-agent');
 var HttpProxyAgent = require('http-proxy-agent');
 var HttpsProxyAgent = require('https-proxy-agent');
 var SocksProxyAgent = require('socks-proxy-agent');
@@ -52,10 +51,6 @@
 exports.proxies.socks5 = SocksProxyAgent;
 exports.proxies.socks5h = SocksProxyAgent;
 
-PacProxyAgent.protocols.forEach(function (protocol) {
-  exports.proxies['pac+' + protocol] = PacProxyAgent;
-});
-
 function httpOrHttps(opts, secureEndpoint) {
   if (secureEndpoint) {
     // HTTPS
